### PR TITLE
Performance test for large state and stable performance

### LINKED ISSUE #12241: Implement JMH benchmark for support process instance creation on larger state
_related to https://github.com/camunda/zeebe/issues/12033_

**Description**

In the [POC](https://github.com/camunda/zeebe/issues/12033#issuecomment-1493803549)  for determining whether the found issues can help us to resolve the issue, I created a simple unit test which also showed us the performance regression on a unit test level.

The test is on a [branch](https://github.com/camunda/zeebe/tree/zell-engine-test), you had to run it in a loop to see the performance differences, to allow jit kick in, etc.

Ideally, we have a [JMH benchmark ](https://github.com/openjdk/jmh) which allows verifying the performance of new process instance creations when we have a larger state or have a similar setup (maybe with decreasing the size of RocksDB which mimics the large state scenarios).

Later, we should think about how we can automate such JMH benchmarks, to avoid regressions. For the first iteration, it should be enough to build such a JMH setup and use it to validate certain changes we plan with #12033 

**Resources:**

 * [integrate JMH with unit tests](https://www.retit.de/continuous-benchmarking-with-jmh-and-junit-2/)
 * [all POC changes in order to resolve the large state issue](https://github.com/camunda/zeebe/compare/rs-stop-iterator-2)
 * [running jmh from eclipse](https://vbochenin.github.io/running-jmh-from-eclipse) 
 * [idea jmh plugin](https://github.com/artyushov/idea-jmh-plugin) pretty cool to easily run the jmh benchmark

### PR DESCRIPTION
**Pull request by ChrisKujawa**
## Description

@oleschoenburg I created this PR in order to get the first increment merged and to not overwhelm you with more changes. Most of the changes here are refactorings and preparation for the JMH test, plus of course the JMH test itself.

 I applied already several things we have discussed, like creation of large state in setup phase, keeping the test in engine module etc. In an upcoming PR I will execute the JMH benchmark inside a unit test, which is executed in a separate CI job, please see [related comment](https://github.com/camunda/zeebe/issues/12241#issuecomment-1591017021).

## Related issues

related to #12241

### CODE REVIEW

**Comment by Zelldon** [comment]
Example execution with `main`

```
 [] INFO  io.camunda.zeebe.engine.perf.EngineLargeStatePerformanceTest - Started 250443 process instances

Result "io.camunda.zeebe.engine.perf.EngineLargeStatePerformanceTest.measureProcessExecutionTime":
  824.767 ±(99.9%) 43.550 ops/s [Average]
  (min, avg, max) = (527.173, 824.767, 948.155), stdev = 87.973
  CI (99.9%): [781.217, 868.317] (assumes normal distribution)


# Run complete. Total time: 00:03:42

Benchmark                                                     Mode  Cnt    Score    Error  Units
EngineLargeStatePerformanceTest.measureProcessExecutionTime  thrpt   50  824.767 ± 43.550  ops/s
```

**Comment by Zelldon** [comment]
Example with SST partitioning disabled
```
[] INFO  io.camunda.zeebe.engine.perf.EngineLargeStatePerformanceTest - Started 209132 process instances

Result "io.camunda.zeebe.engine.perf.EngineLargeStatePerformanceTest.measureProcessExecutionTime":
  150.361 ±(99.9%) 2.380 ops/s [Average]
  (min, avg, max) = (139.649, 150.361, 157.587), stdev = 4.809
  CI (99.9%): [147.980, 152.741] (assumes normal distribution)


Benchmark                                                     Mode  Cnt    Score   Error  Units
EngineLargeStatePerformanceTest.measureProcessExecutionTime  thrpt   50  150.361 ± 2.380  ops/s
```

File: engine/src/test/java/io/camunda/zeebe/engine/perf/TestEngine.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  private final StreamProcessingComposite streamProcessingComposite;
+
+  private TestEngine(
+      final int partitionId, final int partitionCount, final TestContext testContext) {
+
+    final var testStreams =
+        new TestStreams(
+            testContext.temporaryFolder(),
+            testContext.autoCloseableRule(),
+            testContext.actorScheduler());
```
Comment (oleschoenburg) [inline engine/src/test/java/io/camunda/zeebe/engine/perf/TestEngine.java:37]: :thought_balloon: Would be great to get rid of "rules" here. `TestStreams` really doesn't need an `AutoClosableRule` and the `TemporaryFolder` can be replaced by a simple `Path`.

**Review by lenaschoenburg — APPROVED** [review]
Great job @Zelldon :rocket:

Most of my comments are a bit nit-picky and actually unrelated to the JMH benchmark.

:question: I was only able to run these _once_ through IntelliJ, every run after the first is seemingly stuck in the warmup phase. I guess that's not the case for you?

File: engine/src/test/java/io/camunda/zeebe/engine/util/ProcessingExporterTransistor.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+  @Override
+  public void onRecovered(final ReadonlyStreamProcessorContext context) {
+    executorService.submit(
+        () -> {
+          final int partitionId = context.getPartitionId();
+          typedEvent = new TypedRecordImpl(partitionId);
+          final var asyncLogStream = synchronousLogStream.getAsyncLogStream();
+          asyncLogStream.registerRecordAvailableListener(this::onNewEventCommitted);
+          logStreamReader = asyncLogStream.newLogStreamReader().join();
```
[the remark this answers is not in the document]
Comment (Zelldon) [reply engine/src/test/java/io/camunda/zeebe/engine/util/ProcessingExporterTransistor.java:50]: True :+1: just be aware that I just copied/extracted the code :D

File: engine/src/test/java/io/camunda/zeebe/engine/util/ProcessingExporterTransistor.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        });
+  }
+
+  @Override
+  public void onClose() {
+    executorService.shutdownNow();
+  }
+
+  private void onNewEventCommitted() {
+    // this is called from outside (LogStream), so we need to enqueue the task
```
[the remark this answers is not in the document]
Comment (Zelldon) [reply engine/src/test/java/io/camunda/zeebe/engine/util/ProcessingExporterTransistor.java:61]: That is not true. EngineRule and related tests can be run with multiple partitions.

File: engine/src/test/java/io/camunda/zeebe/engine/perf/TestEngine.java (hunk 1 above)
Comment (Zelldon) [reply engine/src/test/java/io/camunda/zeebe/engine/perf/TestEngine.java:37]: Yeah I wanted to stop at some point with the refactoring, I see this mostly in #10004

**Pull request merged**